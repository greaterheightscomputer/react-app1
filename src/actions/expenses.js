import uuid from 'uuid';
import database from '../firebase/firebase';

//Application life cycle in synchrous redux action          //Application life cycle in asynchrous redux action
// component calls action generator                         component calls action generator
// action generator returns object                          action generator returns function
// component dispatches object                              component dispatches function (?)
// redux store changes by running the reducer               function runs (has the ability to dispatch other actions and do whatever it wants)

//4. CREATING THE FOLLOWING ACTION GENETOR FUNCTIONS
// ADD_EXPENSE  
//Synchronous Action Generator Function   
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

//Asynchronous Action Generator Function
export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description='', 
            note='', 
            amount=0, 
            createdAt=0
        } = expenseData; //destructuring from expenseData
        const expense = { description, note, amount, createdAt };
        
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

// REMOVE_EXPENSE    
export const removeExpense = ( { id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

// EDIT_EXPENSE  
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

// //SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

// // asynchronous redux action function
export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];
            // console.log(snapshot.val());
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            }); 

            dispatch(setExpenses(expenses));
        });
    };
};