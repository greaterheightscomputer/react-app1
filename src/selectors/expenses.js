import moment from 'moment';

// 6. Get visible expenses -> this function will match the expenses object with the filters object
// It take in array expenses and destructured filter object and reture filtered and sorted array

export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {  
        // const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        // const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) =>{    //sort() get called on array
          if(sortBy === 'date') {
              return a.createdAt < b.createdAt ? 1 : -1;
        }
        if(sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
      }
    });
};