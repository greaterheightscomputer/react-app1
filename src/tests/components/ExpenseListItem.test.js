import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

test('should render ExpenseListItem with expense', () =>{
    const expense={
        id: expenses[2].id,
        description: expenses[2].description,
        amount: expenses[2].amount,
        createdAt: expenses[2].createdAt,
        note: expenses[2].note
}
    const wrapper = shallow(<ExpenseListItem expenses={expense}/>);
    expect(wrapper).toMatchSnapshot();
});