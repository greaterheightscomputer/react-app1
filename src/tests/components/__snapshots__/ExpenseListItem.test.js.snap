// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with expense 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/edit/undefined"
>
  <div>
    <h3
      className="list-tem__title"
    />
    <span
      className="list-tem__sub-title"
    >
      January 1st, 1970
    </span>
  </div>
  <h3
    className="list-tem__amount"
  >
    â‚¦0.00
  </h3>
</Link>
`;
